scripts/shell.sh

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env sh
+
+# Assumption #1: non-GNU, non-BSD linux distros are out-of-scope
+# Otherwise we at least need to get rid of 'local' keyword
+# Assumption #2: coreutils are present in $PATH
+# Assumption #3: there is internet access (assumption to be removed)
+
+# '-o pipefail' is not POSIX compliant, will fail in some sh's
+set -eu
+
+# Find current script location
+{
+  __FILE__=$(dirname "$(readlink -f $0)")
+} || {
+  echo "Unable to determine current script location.."
+  echo "Will assume '.'"
+  __FILE__="."
+}
+
+fail() {
+  echo "$0:" "$@" >&2
+  exit 1
+}
+
+# Nicked from https://releases.nixos.org/nix/nix-2.5.1/install
+require_util() {
+  command -v "$1" > /dev/null 2>&1 ||
+    fail "you do not have '$1' installed, which I need to $2"
+}
+
+readonly USER_HOME="${HOME:-"HOME variable has not been set"}"
+readonly CACHE_ROOT="${__FILE__}/../.cache"
+readonly NIX_USER_CHROOT_VERSION="1.2.2"
+readonly NIX_USER_CHROOT_DIR="${CACHE_ROOT}/nix-user-chroot"
+readonly NIX_USER_CHROOT_BIN="${NIX_USER_CHROOT_DIR}/nix-user-chroot"
+readonly NIX_STORE="${CACHE_ROOT}/nix-store"
+readonly NIX_VERSION="2.5.1"
+readonly NIX_INSTALL_SCRIPT="${CACHE_ROOT}/nix-${NIX_VERSION}-install.sh"
+readonly NIX_EXTRA_CONF_PATH="${__FILE__}/nix.conf"
+readonly NIX_DIRENV_CONF_PATH="${CACHE_ROOT}/direnv.toml"
+
+IS_NIX_INSTALLED=false
+
+# Check if running on NixOS
+if [ -e "/etc/os-release" ]; then
+  case "$(cat /etc/os-release)" in 
+    *NixOS*)
+      readonly IS_NIXOS=true;
+    ;;
+    *)
+      readonly IS_NIXOS=false;
+    ;;
+  esac
+else
+  readonly IS_NIXOS=false;
+fi
+
+preflightCheck() {
+  require_util curl "download dependencies"
+  require_util tar "decompress nix installation package"
+  require_util unshare "verify kernel support for user namespaces"
+}
+
+setup_nix_user_chroot() {
+  # Verify that user namespaces for unpriviledge users are enabled
+  local readonly are_user_ns_enabled="$(unshare\
```

Comment:
Yeah, definitely won't fly with OSX, so we might need completely separate logic to handle it.

Reply:
Yeah, the root-less approach might not be worth while on OSX, but then we can work on root-supported nix setup, which will work just fine. As for rootless, we would need someone with OSX machine, or an instance on AWS/GCP to make some tests happen.

Reply:
I would propose to leave that decision for PR which implements MacOS support